### ISSUE TITLE
redirect is incorrect when product type not default

### ISSUE THREAD

**Issue by proseLA** [zencart/zencart#3873]
https://github.com/zencart/zencart/blob/dd3875ab3457523e0be7853ff1c75a63492ac27d/admin/product.php#L36-L42

perhaps, this is an edge case scenario.  but i am trying to work within the confines of ZC.

let's say you attempt to create a new product, different from the default, that gets caught in this error.  on the redirect, you would lose the product type, and hence be in a different `type_handler` for the `new_product` action. (i have not tested it with the meta tags preview error; as i find it hard to imagine a separate meta tags collection based on the `type_handler`; but i could be wrong.  and the error would exist there as well.)  

in my testing, a viable solution would be to change the action, and use the messageStack to add the message rather than add session.  my testing shows that changing lines 40-41 to: 

```
          $messageStack->add(ERROR_NO_DATA_TO_SAVE, 'error');
          $action = "new_product";
          break;
```

would now, show the message on this load; as well as not do the redirect and lose the type handler.  i have no problem doing a PR for this change (i can include the `new_product_meta_tags` as well); but i was curious about thoughts prior to proceeding.

**Comment by mc12345678** [comment]
Something else I noticed in the existing redirect is that even if none of the "necessary" information is entered, that anything that has been entered is lost on the redirect, so it would seem that a solution that doesn't lose posted information such as this may be less aggravating, even if the individual repetitively pressed the "preview" button as it floats in the one location (at least on my screen it does I haven't verified CSS has been updated).

It would seem that by allowing the rest of execution to occur outside of this `select`/`case` condition that the display would be loaded/reloaded as if originally entering the `new_product` editor but retaining any other existing entered information.

BTW, to see this in action, create a new Document General product scroll down the screen and note the "missing" fields, alter the product's status away from whatever was shown on load, then without entering any other information select preview.  The error message is displayed at the top, but now when scrolling down, all of the standard product fields are displayed (including price which was otherwise hidden by default (a completely separate issue which has removed some previously convenient workarounds)).

I'm on board with the modification as suggested (though maybe using single quotes instead of double. :P)

**Comment by proseLA** [comment]
i think this solution is correct.  if others agree, i feel this line needs to change as well:

https://github.com/zencart/zencart/blob/3ef5f58ebeaf6505985e66d198b10719b9468432/admin/product.php#L15

unless, i am mistaken, `$param1`, or in this case, `$action` is not changeable by this notifier.  if one wants to do additional validation on `$_POST` data, and  change the action from `new_product_preview` to `new_product` so as to not lose all of the post data, this seems to be the place to do it.

if there is a way to modify `$param1`, ie `$action`, without modifying said code, by all means, let me know.

**Comment by mc12345678** [comment]
There's two things that come to mind without making a modification to the notifier but instead to use what is "generally" available at that point.

One is to capture the `$_POST` data, perform the redirect and then repopulate the `$_POST` data upon coming back through on that redirect.

The other is to reference `$action` as a global variable in the observer, modify it and exit.  This is because generally speaking since this code is executed in the global space (unless the entirety of the Zen Cart store is captured within say a function or method and variables such as `$orders`, `$shipping`, `$payment`, and `$db` are made global so that they could be accessed via all of the other functions and methods that reference such items) the variables within that code section are available.  That said, I could see where there is a goal to have notifiers written in such a way that they perform/support the anticipated action regardless of the "environment" which would allow them to be moved around in future releases and still support the operation(s) originally considered.

Personally, considering the possibility of multiple observers performing operation on the notifier I'm more inclined to perform the redirect, but also if that is/was the standard means of taking action, then the other observer may take issue with the absence of `$_POST` information and itself perform a redirect if it followed the same pattern. Alternatively, if that other observer would take different action based on the value of `$action` had it been supplied as suggested to be `new_product`, then it may miss the opportunity to perform that operation if `$action` is modified "mid-stream" of processing... 

With those two conflicting concepts, that somewhat leads me back to the need to capture `$_POST` data and perform the redirect, but instead of repopulating `$_POST` at the notifier/observer point to instead "inject" it earlier in processing so that it is available for "all" of the associated processing of `$_POST` data...  

That's perhaps a more "comprehensive" approach rather than addressing the single issue discussed above.

Alternatively as a third option, the notifier could be updated to:
````php
$zco_notifier->notify('NOTIFY_BEGIN_ADMIN_PRODUCTS', $action, $action);
````
Allowing the 2nd instance of `$action` (which would be the fourth parameter of the observer's method) to be modifiable from within the observer for this supposedly 'single' situation.

**Comment by mc12345678** [comment]
From the looks of things though, the change that was earlier implemented:
````diff
- $product_type = (isset($_POST['products_id']) ? zen_get_products_type($_POST['products_id']) : isset($_GET['product_type']) ? $_GET['product_type'] : 1);
+ $product_type = (isset($_POST['product_type']) ? $_POST['product_type'] : (isset($_GET['pID']) ? zen_get_products_type($_GET['pID']) : 1));
````
May need to be revisited.  Previously links were generated to include `product_type` in the uri which was "beneficial" in a number of ways; however, in the redesign of the product "navigation" it was basically removed from the uris as it wasn't being used with the above code change effectively being the impetus for "ensuring" it was removed elsewhere.

Though that wouldn't specifically resolve a "loss of `$_POST` data" situation.

**Comment by mc12345678** [comment]
Seems I browser is not cooperating to allow edit of my last message. Wanted to add to the end:
> Though that wouldn't specifically resolve a "loss of `$_POST` data" situation

and require revision of links within at least the `collect_info`, `preview_info` and associated `category_product_listing` new product button to use `get` instead of `post` in: https://github.com/zencart/zencart/blob/b48b65fc4628842394423d3dcf2616c4435017ba/admin/category_product_listing.php#L1027

To be the following:
````diff
-                <?php echo zen_draw_form('newproduct', FILENAME_PRODUCT, 'action=new_product', 'post', 'class="form-horizontal"'); ?>
+               <?php echo zen_draw_form('newproduct', FILENAME_PRODUCT, 'action=new_product', 'get', 'class="form-horizontal"'); ?> 
````


Lastly, thought it would be a good idea to flag this issue up to @Zen4All.

**Comment by mc12345678** [comment]
Seems I browser is not cooperating to allow edit of my last message. Wanted to add to the end:
> Though that wouldn't specifically resolve a "loss of `$_POST` data" situation

and require revision of links within at least the `collect_info`, `preview_info` and associated `category_product_listing` new product button to use `get` instead of `post` in: https://github.com/zencart/zencart/blob/b48b65fc4628842394423d3dcf2616c4435017ba/admin/category_product_listing.php#L1027

To be the following:
````diff
-                <?php echo zen_draw_form('newproduct', FILENAME_PRODUCT, 'action=new_product', 'post', 'class="form-horizontal"'); ?>
+               <?php echo zen_draw_form('newproduct', FILENAME_PRODUCT, 'action=new_product', 'get', 'class="form-horizontal"'); ?> 
````


Lastly, thought it would be a good idea to flag this issue up to @Zen4All.

**Comment by mc12345678** [comment]
Seems I browser is not cooperating to allow edit of my last message. Wanted to add to the end:
> Though that wouldn't specifically resolve a "loss of `$_POST` data" situation

and require revision of links within at least the `collect_info`, `preview_info` and associated `category_product_listing` new product button to use `get` instead of `post` in: https://github.com/zencart/zencart/blob/b48b65fc4628842394423d3dcf2616c4435017ba/admin/category_product_listing.php#L1027

To be the following:
````diff
-                <?php echo zen_draw_form('newproduct', FILENAME_PRODUCT, 'action=new_product', 'post', 'class="form-horizontal"'); ?>
+               <?php echo zen_draw_form('newproduct', FILENAME_PRODUCT, 'action=new_product', 'get', 'class="form-horizontal"'); ?> 
````


Lastly, thought it would be a good idea to flag this issue up to @Zen4All.

**Comment by proseLA** [comment]
@mc12345678 thanks for your input.  not withstanding the fact that it is easiest for me, i think changing the notifier params makes the most sense to me.

**Comment by mc12345678** [comment]
To me it seems that the need to change `$action` is limited and thus if `$action` is to be changed "mid-stream" of processing the page then within the observer at the appropriate point when it is to be altered one could use something like below:

````php
function updateNotifyBeginAdminProducts(&$callingClass, $notifier, $param1) {
    // Some level of code evaluating the information available
    if ($param1 == 'new_product_preview' && $need_to_change) {
        global $action;
        $action = 'new_product';
        return;
    }
    // Other code to complete the observer
}
````
Or:
````php
function updateNotifyBeginAdminProducts(&$callingClass, $notifier, $param1) {
    // Some level of code evaluating the information available
    if ($param1 == 'new_product_preview' && $need_to_change) {
        $GLOBALS['action'] = 'new_product';
        return;
    }
    // Other code to complete the observer
}
````
Where `$need_to_change` is a variable/test identifying that such a "redirect" was needed...

**Comment by drbyte** [comment]
Closed by #3882
